File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/null_object/lab/OutputExpressionVisitor.java
// OutputExpressionVisitor.java

import java.io.PrintWriter;

public class OutputExpressionVisitor extends ExpressionVisitor
{
	public OutputExpressionVisitor(PrintWriter aPrintWriter)
	{
		myPrintWriter = aPrintWriter;
	}

	public void visit(BinaryMinus aBinaryMinus)
	{
		outputInfix(aBinaryMinus, "-");
	}

	public void visit(BinaryPlus aBinaryPlus)
	{
		outputInfix(aBinaryPlus, "+");
	}

	public void visit(Conditional aConditional)
	{
		outputInfix(aConditional, "?", ":");
	}

	public void visit(Divide aDivide)
	{
		outputInfix(aDivide, "/");
	}

	public void visit(Literal aLiteral)
	{
		myPrintWriter.print(aLiteral.getValue());
	}

	public void visit(MethodInvocation aMethodInvocation)
	{
		myPrintWriter.print(aMethodInvocation.getName());
		myPrintWriter.print("(");
		boolean isFirst = true;
		DpIterator<Expression> aIterator = aMethodInvocation.getOperands();
		while (!aIterator.isDone())
		{
			if (isFirst)
			{
				isFirst = false;
			}
			else
			{
				myPrintWriter.print(",");
			}
			Expression aOperand1 = aIterator.getCurrent();
			if (aOperand1 == null)
			{
				throw new DpException("void OutputExpressionVisitor.visit(MethodInvocation aMethodInvocation) -- aOperand1 == null.");
			}
			aOperand1.accept(this);
			aIterator.advance();
		}
		myPrintWriter.print(")");
	}

	public void visit(Multiply aMultiply)
	{
		outputInfix(aMultiply, "*");
	}

	public void visit(UnaryMinus aUnaryMinus)
	{
		outputPrefix(aUnaryMinus, "-");
	}

	public void visit(UnaryPlus aUnaryPlus)
	{
		outputPrefix(aUnaryPlus, "+");
	}

	public void visit(Variable aVariable)
	{
		myPrintWriter.print(aVariable.getName());
	}

	private void outputPrefix(Unary aUnary, String aOperator)
	{
		myPrintWriter.print("(");
		myPrintWriter.print(aOperator);
		if (aUnary.getOperand1() == null)
		{
			throw new DpException("void OutputExpressionVisitor.outputPrefix(Unary aUnary, String aOperator) -- aUnary.getOperand1() == null.");
		}
		aUnary.getOperand1().accept(this);
		myPrintWriter.print(")");
	}

	private void outputPostfix(Unary aUnary, String aOperator)
	{
		myPrintWriter.print("(");
		if (aUnary.getOperand1() == null)
		{
			throw new DpException("void OutputExpressionVisitor.outputPostfix(Unary aUnary, String aOperator) -- aUnary.getOperand1() == null.");
		}
		aUnary.getOperand1().accept(this);
		myPrintWriter.print(aOperator);
		myPrintWriter.print(")");
	}

	private void outputInfix(Binary aBinary, String aOperator)
	{
		myPrintWriter.print("(");
		if (aBinary.getOperand1() == null)
		{
			throw new DpException("void OutputExpressionVisitor.outputInfix(Binary aBinary, String aOperator) -- aBinary.getOperand1() == null.");
		}
		aBinary.getOperand1().accept(this);
		myPrintWriter.print(aOperator);
		if (aBinary.getOperand2() == null)
		{
			throw new DpException("void OutputExpressionVisitor.outputInfix(Binary aBinary, String aOperator) -- aBinary.getOperand2() == null.");
		}
		aBinary.getOperand2().accept(this);
		myPrintWriter.print(")");
	}

	private void outputInfix(Ternary aTernary, String aOperator1, String aOperator2)
	{
		myPrintWriter.print("(");
		if (aTernary.getOperand1() == null)
		{
			throw new DpException("void OutputExpressionVisitor.outputInfix(Ternary aTernary, String aOperator1, String aOperator2) -- aTernary.getOperand1() == null.");
		}
		aTernary.getOperand1().accept(this);
		myPrintWriter.print(aOperator1);
		if (aTernary.getOperand2() == null)
		{
			throw new DpException("void OutputExpressionVisitor.outputInfix(Ternary aTernary, String aOperator1, String aOperator2) -- aTernary.getOperand2() == null.");
		}
		aTernary.getOperand2().accept(this);
		myPrintWriter.print(aOperator2);
		if (aTernary.getOperand3() == null)
		{
			throw new DpException("void OutputExpressionVisitor.outputInfix(Ternary aTernary, String aOperator1, String aOperator2) -- aTernary.getOperand3() == null.");
		}
		aTernary.getOperand3().accept(this);
		myPrintWriter.print(")");
	}

	private PrintWriter myPrintWriter;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/null_object/lab/OutputExpressionVisitor.m
% OutputExpressionVisitor.m

classdef OutputExpressionVisitor < ExpressionVisitor
	methods
		function visitBinaryMinus(theOutputExpressionVisitor, aBinaryMinus)
			theOutputExpressionVisitor.outputInfix2(aBinaryMinus, '-');
		end

		function visitBinaryPlus(theOutputExpressionVisitor, aBinaryPlus)
			theOutputExpressionVisitor.outputInfix2(aBinaryPlus, '+');
		end

		function visitConditional(theOutputExpressionVisitor, aConditional)
			theOutputExpressionVisitor.outputInfix3(aConditional, '?', ':');
		end

		function visitDivide(theOutputExpressionVisitor, aDivide)
			theOutputExpressionVisitor.outputInfix2(aDivide, '/');
		end

		function visitLiteral(theOutputExpressionVisitor, aLiteral)
			fprintf('%d', aLiteral.value);
		end

		function visitMethodInvocation(theOutputExpressionVisitor, aMethodInvocation)
			fprintf('%s', aMethodInvocation.Name);
			fprintf('(');
			isFirst = true;
			aIterator = aMethodInvocation.getOperands;
			while ~aIterator.isDone
				if isFirst
					isFirst = false;
				else
					fprintf(',');
				end
				aOperand = aIterator.getCurrent;
				if isempty(aOperand)
					error('aOperand is empty.');
				end
				aOperand.accept(theOutputExpressionVisitor);
				aIterator.advance;
			end
			fprintf(')');
		end

		function visitMultiply(theOutputExpressionVisitor, aMultiply)
			theOutputExpressionVisitor.outputInfix2(aMultiply, '*');
		end

		function visitUnaryMinus(theOutputExpressionVisitor, aUnaryMinus)
			theOutputExpressionVisitor.outputPrefix(aUnaryMinus, '-');
		end

		function visitUnaryPlus(theOutputExpressionVisitor, aUnaryPlus)
			theOutputExpressionVisitor.outputPrefix(aUnaryPlus, '+');
		end

		function visitVariable(theOutputExpressionVisitor, aVariable)
			fprintf('%s', aVariable.Name);
		end
	end

	methods (Access = private)
		function outputPrefix(theOutputExpressionVisitor, aUnary, aOperator)
			fprintf('(');
			fprintf('%s', aOperator);
			if isempty(aUnary.Operand1)
				error('Operand1 is empty.');
			end
			aUnary.Operand1.accept(theOutputExpressionVisitor);
			fprintf(')');
		end

		function outputPostfix(theOutputExpressionVisitor, aUnary, aOperator)
			fprintf('(');
			if isempty(aUnary.Operand1)
				error('Operand1 is empty.');
			end
			aUnary.Operand1.accept(theOutputExpressionVisitor);
			fprintf('%s', aOperator);
			fprintf(')');
		end

		function outputInfix2(theOutputExpressionVisitor, aBinary, aOperator)
			fprintf('(');
			if isempty(aBinary.Operand1)
				error('Operand1 is empty.');
			end
			aBinary.Operand1.accept(theOutputExpressionVisitor);
			fprintf('%s', aOperator);
			if isempty(aBinary.Operand2)
				error('Operand2 is empty.');
			end
			aBinary.Operand2.accept(theOutputExpressionVisitor);
			fprintf(')');
		end

		function outputInfix3(theOutputExpressionVisitor, aTernary, aOperator1, aOperator2)
			fprintf('(');
			if isempty(aTernary.Operand1)
				error('Operand1 is empty.');
			end
			aTernary.Operand1.accept(theOutputExpressionVisitor);
			fprintf('%s', aOperator1);
			if isempty(aTernary.Operand2)
				error('Operand2 is empty.');
			end
			aTernary.Operand2.accept(theOutputExpressionVisitor);
			fprintf('%s', aOperator2);
			if isempty(aTernary.Operand3)
				error('Operand3 is empty.');
			end
			aTernary.Operand3.accept(theOutputExpressionVisitor);
			fprintf(')');
		end
	end
end
